File: Programmers/Lv2/largest_number.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool cmp(string a, string b){
    return a+b>b+a;
}

string solution(vector<int> numbers) {
    string answer = "";
    vector<string> num_str;
    
    for(auto num : numbers)
        num_str.push_back(to_string(num));
    
    sort(num_str.begin(),num_str.end(),cmp);
    
    if(num_str.at(0) =="0") return "0";
    for(auto c : num_str)
        answer+=c;
    
    return answer;
}

File: java/src/programmers/lv2/largest_number.java
package programmers.lv2;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.*;

public class largest_number {
    public String solution(int[] numbers) {
        String answer;
        Queue<String> q = new PriorityQueue<>((String a,String b) -> (b+a).compareTo(a+b));
        Arrays.stream(numbers).forEach( n -> q.offer(Integer.toString(n)));

        StringBuilder sb = new StringBuilder();
        while(!q.isEmpty()){
            sb.append(q.peek());
            q.poll();
        }
        answer=sb.toString();
        return answer.charAt(0) == '0' ? "0" : answer;
    }
    @Test
    void test(){
        Assertions.assertEquals("6210", solution(new int[]{6,10,2}));
        Assertions.assertEquals("9534330", solution(new int[]{3, 30, 34, 5, 9}));
        Assertions.assertEquals("0", solution(new int[]{0,0,0,0,0}));
    }
}
